From 225c523c03f9b4849f87281ed12f177c6131f567 Mon Sep 17 00:00:00 2001
From: Zhang Zekun <zhangzekun11@huawei.com>
Date: Fri, 16 Dec 2022 07:21:44 +0000
Subject: [PATCH 4/4] iommu: Check the iova range in the iommu_map

-----------------------------------------------

During iommu mapping, check whether the iova range is in the
specified domain geometry.

Signed-off-by: Zhang Zekun <zhangzekun11@huawei.com>
Signed-off-by: Zhou Guanghui <zhouguanghui1@huawei.com>
---
 drivers/iommu/iommu.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/iommu/iommu.c b/drivers/iommu/iommu.c
index 97953fa27..f80596080 100644
--- a/drivers/iommu/iommu.c
+++ b/drivers/iommu/iommu.c
@@ -2613,6 +2613,11 @@ static int __iommu_map(struct iommu_domain *domain, unsigned long iova,
 	if (unlikely(!(domain->type & __IOMMU_DOMAIN_PAGING)))
 		return -EINVAL;
 
+	if (domain->geometry.force_aperture &&
+	    (iova < domain->geometry.aperture_start ||
+	     (iova + size - 1) > domain->geometry.aperture_end))
+		return -ERANGE;
+
 	/* find out the minimum page size supported */
 	min_pagesz = 1 << __ffs(domain->pgsize_bitmap);
 
-- 
2.17.1

